### ISSUE TITLE
[GlobalStep - WooCommerce 4.4 RC 1] "Filter Product by Attribute" block not working as expected.

### ISSUE THREAD

**Issue by gglobalstep** [woocommerce/woocommerce-gutenberg-products-block#2971]
## Bug Description:
"Filter Product by Attribute" block not working as expected

## Environment:
**Windows 10, MAC**
**Browser:** Chrome(Version 83.0.4103.61 (64-bit))
                 Firefox(Version 77.0.1)(64-bit)
                 Safari(Version 13.1.1)

## Steps to Reproduce:
1. Create a Jurassic Ninja site.
1. Install and activate all the required plugins.
1. Create two variable products say Big Shoes and Small Shoes.
1. Add color attribute having values Pink, Green, Blue, Black, Brown.
1. Create Big shoes with all the variations.
1. Create Small Shoes with any three variations say Black, Blue, Brown.
1. Publish both the products.
1. Go to Pages.
1. Add New page.
1. Add the "All Products" block.
1. Add "Filter Product by Attribute"
1. Observe that we are able to see all the products(i.e. Small Shoes and Big Shoes) in all the 5 variations(i.e. Black, Blue, pink, green, brown).

## Actual Result:
"Filter Product by Attribute" block not working as expected

## Expected Result:
"Filter Product by Attribute" block should work as expected.

## Artifacts:
[#2971_Product_Export_driving_eel.zip](https://github.com/woocommerce/woocommerce-gutenberg-products-block/files/5029200/2971_Product_Export_driving_eel.zip)

**Isolating the problem (mark completed items with an [x]):**
- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

<details>
```
`
### WordPress Environment ###

WordPress address (URL): https://driving-eel.jurassic.ninja
Site address (URL): https://driving-eel.jurassic.ninja
WC Version: 4.4.0
REST API Version: ✔ 1.0.11
WC Blocks Version: ✔ 3.1.0
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 1.4.0-beta.3
Log Directory Writable: ✔
WP Version: 5.5-RC2-48740
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.43 (Unix) OpenSSL/1.0.2g
PHP Version: 7.4.8
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.31-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 4.3.1
WC Database Prefix: wp_
Total Database Size: 3.91MB
Database Data Size: 2.45MB
Database Index Size: 1.46MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_mailchimp_carts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_mailchimp_jobs: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_options: Data: 1.48MB + Index: 0.08MB + Engine InnoDB
wp_postmeta: Data: 0.08MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
customize_changeset: 1
page: 8
post: 2
product: 14
product_variation: 20
revision: 5

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (11) ###

Query Monitor: by John Blackbourn – 3.6.1
Companion Plugin: by Osk – 1.17
Facebook for WooCommerce: by Facebook – 2.0.0 – Installed version not tested with active version of WooCommerce 4.4.0
Jetpack by WordPress.com: by Automattic – 8.8
Google Ads for WooCommerce: by Kliken – 1.0.4 – Installed version not tested with active version of WooCommerce 4.4.0
Mailchimp for WooCommerce: by Mailchimp – 2.4.3 – Installed version not tested with active version of WooCommerce 4.4.0
WooCommerce Beta Tester: by WooCommerce – 2.0.1 – Installed version not tested with active version of WooCommerce 4.4.0
WooCommerce Payments: by Automattic – 1.2.0 – Installed version not tested with active version of WooCommerce 4.4.0
WooCommerce Services: by Automattic – 1.24.0 – Installed version not tested with active version of WooCommerce 4.4.0
WooCommerce: by Automattic – 4.4.0-rc.1
WordPress Beta Tester: by Peter Westwood
Andy Fragen – 2.2.11


### Inactive Plugins (2) ###

Akismet Anti-Spam: by Automattic – 4.1.6
Hello Dolly: by Matt Mullenweg – 1.7.2

### Dropin Plugins (1) ###

db.php: Query Monitor Database Class

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 2.5.8
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Canceled: 1
Oldest: 2020-08-05 11:01:58 +0000
Newest: 2020-08-05 11:01:58 +0000

Complete: 4
Oldest: 2020-08-05 11:01:58 +0000
Newest: 2020-08-05 11:07:42 +0000

`
```
</details>

**Comment by nerrad** [comment]
Related to work done in https://github.com/woocommerce/woocommerce/pull/26260 which fixes the issue in core but does not address the same thing happening in our blocks.

**Comment by nerrad** [comment]
I can reproduce this with the latest version of Woo Blocks against both Woo Core 4.4.rc1 and Woo Core 4.3.1. 

It looks like this is a pre-existing issue but I've assigned this to a tentative 3.1.1 milestone pending a decision on whether this is fixed for the Woo Core 4.4 release or not.

**Comment by nerrad** [comment]
Actually, this doesn't look like a bug.  By adding _available_ variations to the attributes when creating the product, it doesn't matter what actual product variations are _setup_ because from the perspective of any query, the main product has all variations available and that's how it is saved in the database.

For instance, if I setup a "Small Shoes" product and add variations to attributes like this:

<img width="1433" alt="Image 2020-08-05 at 3 24 50 PM" src="https://user-images.githubusercontent.com/1429108/89455166-dda16c80-d72f-11ea-965b-e05b111dcae9.png">

And in the _variations tab..._

<img width="1423" alt="Image 2020-08-05 at 3 25 45 PM" src="https://user-images.githubusercontent.com/1429108/89455232-f873e100-d72f-11ea-86f4-7b66b95e4174.png">

Then there are three related products created. The main one (`product` post_type) which has term relationships to all the variation colors for the color attribute. And the "Gray" and "Red" product variations (`product_variation` post type), one product variation attached to the Gray term (via term_relationships) and one attached to the Red term (via term relationships). The current query pulls from `product` post type and thus is considering _all the variations indicated are available for that product_. If a product is only supposed to be restricted to _actual_ configured variations, then shouldn't that be configured at the product level too?

So this seems like a bit of a ui/ux conundrum too.

cc @woocommerce/proton because while I see where the query we're doing in Blocks can be modified to only query against `product_variation` product types and then get the parent id for what products to return, there seems to be a ui/ux issue that could be improved here too? Also hoping you can educate me if I'm missing some context here.

So on the query,
